testsuite/shared/src/main/java/org/jboss/as/test/integration/transactions/TransactionCheckerSingleton.java

```diff
@@ -40,8 +40,8 @@
 @Remote
 @ManagedBean
 public class TransactionCheckerSingleton implements TransactionCheckerSingletonRemote {
-    private int committed, prepared, rolledback;
-    private int  synchronizedBegin, synchronizedBefore, synchronizedAfter,
+    private volatile int committed, prepared, rolledback;
+    private volatile int synchronizedBegin, synchronizedBefore, synchronizedAfter,
```

Comment:
While making these volatile might improve the situation, it will not solve it as the ++ unary operator is not atomic.  I would suggest changing these to use AtomicInteger.

Reply:
Since this is a singleton ejb with container-managed concurrency, all method access are serialized and no two methods will be accessing the instance state. So I think make them volatile should work. If you agree, we can add volatile to `messages` field as well.

Reply:
OK - in that case, volatile is sufficient.  `messages` should not need to be volatile since it is effectively final.